src/Build.UnitTests/Graph/ProjectGraph_Tests.cs

```diff
@@ -97,10 +98,42 @@ public void ConstructWithCycle()
                 TransientTestFile entryProject = CreateProject(env, 1, new[] { 2 });
                 CreateProject(env, 2, new[] { 3 });
                 CreateProject(env, 3, new[] { 1 });
+                Should.Throw<CircularDependencyException>(() => new ProjectGraph(entryProject.Path));
+            }
+        }
 
-                // TODO: This should eventually throw, but for now not infinite-looping is sufficient.
-                ProjectGraph graph = new ProjectGraph(entryProject.Path);
-                graph.ProjectNodes.Count.ShouldBe(3);
+        [Fact]
+        public void ConstructWithSelfLoop()
+        {
+            using (var env = TestEnvironment.Create())
+            {
+                TransientTestFile entryProject = CreateProject(env, 1, new[] { 2, 3 });
+                CreateProject(env, 2, new[] { 2 });
+                CreateProject(env, 3);
+                Should.Throw<CircularDependencyException>(() => new ProjectGraph(entryProject.Path));
+            }
+        }
+
+        [Fact]
+        // graph with a cycle between 2->6->7->3->2
+        public void ConstructBigGraphWithCycle()
+        {
+            using (var env = TestEnvironment.Create())
+            {
+                TransientTestFile entryProject = CreateProject(env, 1, new[] {2,3,4});
+                CreateProject(env, 2, new[] {5, 6});
+                CreateProject(env, 3, new[] {2, 8});
+                CreateProject(env, 4);
+                CreateProject(env, 5, new []{9, 10});
+                var proj6 = CreateProject(env, 6, new[] { 7});
+                var proj7 = CreateProject(env, 7, new[] { 3 });
+                CreateProject(env, 8);
+                CreateProject(env, 9);
+                CreateProject(env, 10);
+                string projects = proj6.Path + Environment.NewLine + proj7.Path;
+                // the error message can contain a cycle from 2->3->6->7 or 3->6->7->2
+                // the error message should always contain 6.proj and 7.proj
```

Comment:
Shouldn't it also contain 2 and 3?

Seems like the error should be something like:
```
2.proj
3.proj
7.proj
6.proj
2.proj
```

IE, the first and last line should be the same to complete the cycle.

Reply:
It will contain 2 and 3. Since 2 and 3 are at the same depth, I am not sure if 2 will always be processed first. If 3 gets processed first, then our error message will contain 3->2->6->7->3. Asserting the exact order may result in flakiness